ingenialink/ethercat/servo.py

```diff
@@ -0,0 +1,126 @@
+from typing import Optional
+
+from pysoem import CdefSlave
+import ingenialogger
+
+from ingenialink.exceptions import ILIOError
+from ingenialink.servo import Servo
+from ingenialink.canopen.dictionary import CanopenDictionary
+from ingenialink.canopen.register import CanopenRegister, REG_DTYPE, REG_ACCESS
+from ingenialink.constants import CAN_MAX_WRITE_SIZE
+
+logger = ingenialogger.get_logger(__name__)
+
+
+class EthercatServo(Servo):  # type: ignore
+    """Ethercat Servo instance.
+
+    Args:
+        slave: Slave to be connected.
+        slave_id: Slave ID.
+        dictionary_path: Path to the dictionary.
+        servo_status_listener: Toggle the listener of the servo for
+            its status, errors, faults, etc.
+
+    """
+
+    DICTIONARY_CLASS = CanopenDictionary
+    MAX_WRITE_SIZE = CAN_MAX_WRITE_SIZE
+
+    STATUS_WORD_REGISTERS = "CIA402_DRV_STATE_STATUS"
+    RESTORE_COCO_ALL = "CIA301_COMMS_RESTORE_ALL"
+    STORE_COCO_ALL = "CIA301_COMMS_STORE_ALL"
+    MONITORING_DATA = CanopenRegister(
+        identifier="MONITORING_DATA",
+        units="",
+        subnode=0,
+        idx=0x58B2,
+        subidx=0x01,
+        cyclic="CONFIG",
+        dtype=REG_DTYPE.U16,
+        access=REG_ACCESS.RO,
+    )
+    DIST_DATA = CanopenRegister(
+        identifier="DISTURBANCE_DATA",
+        units="",
+        subnode=0,
+        idx=0x58B4,
+        subidx=0x01,
+        cyclic="CONFIG",
+        dtype=REG_DTYPE.U16,
+        access=REG_ACCESS.WO,
+    )
+
+    def __init__(
+        self,
+        slave: CdefSlave,
+        slave_id: int,
+        dictionary_path: Optional[str] = None,
+        servo_status_listener: bool = False,
+    ):
+        self.__slave = slave
+        self.slave_id = slave_id
+        super(EthercatServo, self).__init__(slave.name, dictionary_path, servo_status_listener)
+
+    def _read_raw(
+        self, reg: CanopenRegister, buffer_size: int = 0, complete_access: bool = False
+    ) -> bytes:
+        self._lock.acquire()
+        try:
+            value: bytes = self.__slave.sdo_read(reg.idx, reg.subidx, buffer_size, complete_access)
```

Comment:
If the wire is disconnected read returns a random value

Reply:
Yes, I saw that. I think that the only thing we can do is to check the slave status and if is not alive, raise an Exception. What do you think?

Reply:
Solved using the working counter (wkc).